fixes #23369 - load vmware datastores async

timogoebel opened the pull request:
This commit introduces async loading of vSphere Datastores and Storage Pods.

This speeds up the time before the form is required and is a prerequisite for #3559, a feature requested in 2013 that we still haven't delivered.

![foreman-vsphere-load-datastores-async-3](https://user-images.githubusercontent.com/4107560/46404134-9ba6da00-c704-11e8-9c40-13d242578a3d.gif)

@ohadlevy: Do you mind reviewing this as this is react code?

timogoebel committed:
fixes #23369 - load vmware datastores async

amirfefer commented on webpack/assets/javascripts/react_app/components/common/forms/Select.js:
```diff
@@ -1,14 +1,29 @@
+    const { allowClear } = this.props;
+
+    if ($.fn.select2) {
+      $(this.refs.select)
```
react 16.3 uses `createRef` API instead
Would you mind to use it?
https://reactjs.org/docs/refs-and-the-dom.html

amirfefer commented on webpack/assets/javascripts/react_app/redux/reducers/hosts/storage/vmware.js:
```diff
@@ -72,7 +78,37 @@ export default (state = initialState, { type, payload }) => {
+        .set('datastoresError', undefined)
+        .set('storagePods', {})
+        .set('storagePodsLoading', true)
+        .set('storagePodsError', undefined)
```
instead of a lot of nested `.set` you can use `merge`

amirfefer commented on webpack/assets/javascripts/react_app/redux/actions/hosts/storage/vmware.js:
```diff
@@ -17,14 +24,38 @@ export const updateDisk = (key, newValues) => ({
+  });
+};
+
+export const fetchDatastores = url => (dispatch) => {
```
should it be exported while it's used only in this file?

amirfefer replied:
on the other hand, i tend to prefer to call this action from the component's [componentDidMount](https://github.com/theforeman/foreman/blob/ae96cb70130968bdcc8b7a9a9af7edc7c6e0d9a8/webpack/assets/javascripts/react_app/components/hosts/storage/vmware/index.js#L26)

amirfefer commented on webpack/assets/javascripts/react_app/redux/actions/hosts/storage/vmware.js:
```diff
@@ -17,14 +24,38 @@ export const updateDisk = (key, newValues) => ({
+  });
+};
+
+export const fetchStoragePods = url => (dispatch) => {
```
ditto

amirfefer commented on webpack/assets/javascripts/react_app/components/hosts/storage/vmware/controller/index.js:
```diff
@@ -17,6 +17,12 @@ const Controller = ({
   controllerVolumes,
   removeController,
   config,
+  datastores,
```
Could you add Proptype?  It would be much easier to understand the component's props with it

amirfefer reviewed (changes requested):
Thanks @timogoebel  :+1: 
I left few inline comments about the javascript part :)

timogoebel commented:
@amirfefer: Thanks for the review, will take a look at your invaluable comments (and probably learn a lot!). 👍

timogoebel commented:
@amirfefer: I hope I addressed most of your comments. Unfortunately, I did not get `createRef` working. I guess there is enough refactoring in this PR already, we can always change that later.

ares commented:
I've tested the functionality and works great with and without cache enabled. I defer to @amirfefer for the `createRef` pattern, but :+1: form me for manual testing. I haven't really read the code. Test failures are unrelated.

ohadlevy commented on webpack/assets/javascripts/react_app/redux/reducers/hosts/storage/vmware.js:
```diff
@@ -67,12 +73,45 @@ export default (state = initialState, { type, payload }) => {
+      return state
+        .set('datastoresError', undefined)
+        .set('datastores', {})
+        .set('datastoresLoading', true);
```
please use merge instead of multiple sets, its suppose to be faster :)

ohadlevy reviewed (commented):
could you please also have a look at the storybook? I think it needs an update after these changes.

nice additions, thanks!

timogoebel commented:
@ohadlevy: Sure, but after 20 minutes trying to get the storybook server running I gave up. Does it work for you?

```
$ npm run storybook

> TheForemanDevDeps@1.20.0 storybook /Users/f0218175/foremandev/foreman
> start-storybook -p 6006

@storybook/react v3.2.13

=> Loading custom .babelrc
=> Loading custom addons config.
=> Loading custom webpack config (full-control mode).
/Users/f0218175/foremandev/foreman/.storybook/webpack.config.js:7
  defaultConfig.module.rules = [
                ^

TypeError: Cannot read property 'module' of undefined
    at module.exports (/Users/f0218175/foremandev/foreman/.storybook/webpack.config.js:7:17)
    at exports.default (/Users/f0218175/foremandev/foreman/node_modules/@storybook/react/dist/server/config.js:62:12)
    at exports.default (/Users/f0218175/foremandev/foreman/node_modules/@storybook/react/dist/server/middleware.js:19:37)
    at Object.<anonymous> (/Users/f0218175/foremandev/foreman/node_modules/@storybook/react/dist/server/index.js:153:34)
    at Module._compile (module.js:660:30)
    at Object.Module._extensions..js (module.js:671:10)
    at Module.load (module.js:573:32)
    at tryModuleLoad (module.js:513:12)
    at Function.Module._load (module.js:505:3)
    at Module.require (module.js:604:17)
```

ohadlevy commented:
@timogoebel it seems to work for me:
```
Switched to a new branch '6103'
➜  foreman git:(6103) npm run storybook                                                                                                                     git:(6103|) 

> TheForemanDevDeps@1.20.0 storybook /home/ohad/git/foreman
> start-storybook -p 6006

info @storybook/react v3.4.8
info 
info => Loading custom .babelrc
info => Loading custom addons config.
info => Loading custom webpack config (full-control mode).
 66% building modules 623/657 modules 34 active ...oreman/node_modules/lodash/isArray.jsDEPRECATION WARNING on line 1, column 8 of stdin:
Including .css files with @import is non-standard behaviour which will be removed in future versions of LibSass.
Use a custom importer to maintain this behaviour. Check your implementations documentation on how to create a custom importer.

webpack built d4fc339334bb701f4e07 in 13336ms                                             
info Storybook started on => http://localhost:6006/
```

ares commented:
I have the similar issue without this PR, the npm command starts the server but navigating to storybook, I see `Cannot read property 'defaults' of undefined` 
![story](https://user-images.githubusercontent.com/109773/46458787-38c24b00-c7b7-11e8-93bc-cd54b86e6c73.png)

amirfefer commented on webpack/assets/javascripts/react_app/components/hosts/storage/vmware/index.js:
```diff
@@ -93,10 +131,43 @@ class StorageContainer extends React.Component {
+  config: PropTypes.object,
+  controllers: PropTypes.array,
+  data: PropTypes.object,
+  datastores: PropTypes.object,
```
PropTypes.object is not informative 
`Proptype.shape` is much better for arrays and objects.
Plus, if a prop is required, add `isRequired` to it
i.e:
```javascript
PropTypes.shape({
    color: PropTypes.string,
    fontSize: PropTypes.number
  }).isRequired, // if a prop is required, add `.isRequired`, else just add a default value
```

amirfefer commented on webpack/assets/javascripts/react_app/components/hosts/storage/vmware/index.js:
```diff
@@ -93,10 +131,43 @@ class StorageContainer extends React.Component {
 const mapDispatchToProps = (state) => {
-  const { controllers, config, volumes = [] } = state.hosts.storage.vmware;
+  const {
+    controllers, config, volumes = [], datastores, datastoresLoading, datastoresError,
```
you can use default values with Proptype:
```javascript
StorageContainer.defaultProps = {
  volumes: []
};
```

amirfefer reviewed (changes requested):
Thanks @timogoebel :+1: 
just few minor issues with `PropType`

ohadlevy commented:
what is the status of this PR? (if its ready for today's cut of date?)

timogoebel commented:
@ohadlevy: I won't be able to look at this today. Let's to it in the 1.21 cycle.

timogoebel commented:
@amirfefer, @ohadlevy, @ares: Improved the prop types and fixed up storybook. Can you please re-review?

amirfefer reviewed (dismissed):
JS part looks good , thanks @timogoebel :+1: 
I did not test it though (have no vmware)

amirfefer commented:
regardless of this PR,  I'd like to mention that i'm not a fan of spinners, especially when it comes on inputs, because it might lead to too many spinners in a page.  There's a much better solution IMO - skeleton loading:
for example:

![skeleton](https://user-images.githubusercontent.com/11807069/47005198-e9c3d080-d13b-11e8-96b6-78b06d9fc004.gif)

@theforeman/ui-ux, @Rohoover  any thoughts ?

timogoebel commented:
@amirfefer: I fully agree, but that's definitely out of the scope of this PR.

terezanovotna commented:
Love the idea of skeleton loading! [Opened a design issue for it. ](https://github.com/patternfly/patternfly-design/issues/726)

Can we make the button next to Disk Name look more like [patternfly button?](https://www.patternfly.org/pattern-library/widgets/#buttons)

timogoebel commented:
> Can we make the button next to Disk Name look more like [patternfly button?](https://www.patternfly.org/pattern-library/widgets/#buttons)

We can, but not in this PR. :-)

Rohoover commented:
I like the skeleton as well but we would need to get this into PF since this is a departure.

timogoebel commented:
@amirfefer, @ares: Anything missing here?

amirfefer commented:
LGFM :+1:

timogoebel commented:
This is my final attempt to get this in, I rebased the code and added support for statistics of the datastore and storage pods.

ohadlevy reviewed (dismissed):
LGTM, I don't have a working VMWare env to test.

lizagilman commented:
@timogoebel or anyone interested, regarding storybook issue (if you still have it), this workaround works for me - https://gist.github.com/lizagilman/d3a94114fc3fd3214138ec120fca4a5c

timogoebel commented:
@lizagilman: Thanks. I chose the "reboot" approach. That was quite popular in ancient times.
`rm -rf node_module && npm install`.

timogoebel commented:
@chris1984: Can you do me a favor and test and ack this? Thanks a bunch.

chris1984 commented:
> @chris1984: Can you do me a favor and test and ack this? Thanks a bunch.

Sure thing :)

amirfefer commented:
I've test it with vmware env and it works for me:
![vmware](https://user-images.githubusercontent.com/11807069/48009154-bf09de00-e123-11e8-8ce3-3e6cba038525.png)

Looks like a snapshot needs an update

chris1984 reviewed (dismissed):
ACK, looks good Thanks @timogoebel

chris1984 commented:
[test foreman]

ares commented:
It seems like test failures are related, mind to take a look?

timogoebel commented:
Fixed the tests.

timogoebel commented:
Tests passed.

ohadlevy commented on package.json:
```diff
@@ -83,6 +83,7 @@
     "jstz": "~1.0.7",
     "lodash": "^4.17.10",
     "multiselect": "~0.9.12",
+    "number_helpers": "^0.1.1",
```
I tried packaging this, but it fails, I assume before merge we would need a packaging ACK.

EDIT: for completion, the error is
```
./add_npm_package.sh number_helpers 0.1.1
Found 0 dependencies - using single strategy
Making directory...FINISHED
Creating specs and downloading sources...---- npm2rpm ----
-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
 - Starting npm module download: https://registry.npmjs.org/number_helpers/-/number_helpers-0.1.1.tgz
 - Unpacking in /tmp/npm2rpm-15831E09pZQE4YSYD ...
 - Finished extracting for number_helpers
 - Reading package.json for number_helpers
Warning: No README data
Warning: No license field.
 - Finished reading package.json for number_helpers
fs.js:133
  throw new ERR_INVALID_CALLBACK();
  ^

TypeError [ERR_INVALID_CALLBACK]: Callback must be a function
    at maybeCallback (fs.js:133:9)
    at Object.writeFile (fs.js:1139:14)
    at writeSpecFile (/home/ohad/git/foreman-packaging/node_modules/npm2rpm/bin/npm2rpm.js:87:6)
    at Extract.tar_extract.stream.on (/home/ohad/git/foreman-packaging/node_modules/npm2rpm/bin/npm2rpm.js:78:5)
    at Extract.emit (events.js:187:15)
    at DirWriter.<anonymous> (/home/ohad/git/foreman-packaging/node_modules/tar/lib/extract.js:81:8)
    at DirWriter.emit (events.js:187:15)
    at end (/home/ohad/git/foreman-packaging/node_modules/fstream/lib/writer.js:324:14)
    at /home/ohad/git/foreman-packaging/node_modules/fstream/lib/writer.js:314:34
    at endUtimes (/home/ohad/git/foreman-packaging/node_modules/fstream/lib/writer.js:238:46)
```

ekohl replied:
I think that was a warning for a while already but likely that there's an update which makes it strict. I'll have a look.

ohadlevy replied:
also TODO: add this package to webpack vendor file (can be done in a follow up pr).

ekohl replied:
https://github.com/theforeman/foreman-packaging/pull/3174. Luckily in https://github.com/dLobatog/npm2rpm/pull/48 I already fixed that issue.

ohadlevy replied:
@ekohl great, once you are ready please ack this PR.

ekohl reviewed (approved):
Packaging ACK, haven't looked at the code :)

ohadlevy commented:
thanks @timogoebel!

timogoebel commented:
Thanks, @ohadlevy + @amirfefer + @ares + @ekohl + @chris1984!
